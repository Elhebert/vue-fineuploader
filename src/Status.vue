<template>
  <renderless-status
    :id="id"
    :uploader="uploader"
  >
    <span slot-scope="{ status }">{{ statuses[status] }}</span>
  </renderless-status>
</template>

<script>
import RenderlessStatus from './renderless/Status'

export default {
  props: {
    id: {
      type: Number,
      required: true,
    },
    uploader: {
      type: Object,
      required: true,
    },
  },

  components: {
    RenderlessStatus,
  },

  data: () => ({
    statuses: {
      deleting: 'Deleting...',
      paused: 'Paused',
      queued: 'Queued',
      retrying_upload: 'Retrying...',
      submitted: 'Submitted',
      submitting: 'Submitting...',
      uploading: 'Uploading...',
      upload_failed: 'Failed',
      upload_successful: 'Completed',
    }
  })
}
</script>
