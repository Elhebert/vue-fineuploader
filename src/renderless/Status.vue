<script>
export default {
  props: {
    id: {
      type: Number,
      required: true,
    },
    uploader: {
      type: Object,
      required: true,
    },
  },

  data: () => ({
    status: null,
    unmounted: false,
  }),

  mounted() {
    this.uploader.on('statusChange', this.onStatusChange)
  },

  beforeDestroy() {
    this.unmounted = true
    this.unregisterStatusChangeHandler()
  },

  methods: {
    onStatusChange(id, oldStatus, newStatus) {
      if (id !== this.id || this.unmounted) {
        return
      }

      this.status = newStatus
    },

    unregisterStatusChangeHandler() {
      this.uploader.off('statusChange', this.onStatusChange)
    },
  },

  render() {
    return this.$scopedSlots.default({
      status: this.status,
    })
  },
}
</script>
