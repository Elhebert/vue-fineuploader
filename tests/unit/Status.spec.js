import { mount } from '@vue/test-utils'
import Status from '@/Status.vue'
import FineUploader from 'fine-uploader-wrappers'
import { status as STATUSES } from 'fine-uploader/lib/core/all'

describe('Status.vue', () => {
  let wrapper
  let uploader
  let renderless

  beforeEach(() => {
    uploader = new FineUploader({ options: {} })

    wrapper = mount(Status, {
      propsData: { id: 0, uploader },
    })
    renderless = wrapper.vm.$children[0]
  })

  it('render nothing for a different file', () => {
    renderless.onStatusChange(1, STATUSES.UPLOAD_SUCCESSFUL, STATUSES.DELETING)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.text()).toBe('')
  })

  it('render nothing for an untracked status value', () => {
    renderless.onStatusChange(0, STATUSES.DELETING, STATUSES.DELETE_FAILED)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.text()).toBe('')
  })

  it('renders the status for the tracked file', () => {
    renderless.onStatusChange(0, STATUSES.UPLOAD_SUCCESSFUL, STATUSES.DELETING)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.text()).toBe('Deleting...')
  })
})
