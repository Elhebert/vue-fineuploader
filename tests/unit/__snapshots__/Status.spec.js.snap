// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status.vue render nothing for a different file 1`] = `<span></span>`;

exports[`Status.vue render nothing for an untracked status value 1`] = `<span></span>`;

exports[`Status.vue renders the status for the tracked file 1`] = `<span>Deleting...</span>`;
